### ISSUE TITLE
tools/flow: cue task not execute, when result in non-task value

### ISSUE THREAD

**Issue by morlay** [cue-lang/cue#2965]
### What version of CUE are you using (`cue version`)?

<pre>
$ cue version
v0.8.0
</pre>

### Does this issue reproduce with the latest stable release?

yes

### What did you do?

```
-- main.go --
package main

import (
	"context"
	"log"
	"os"

	"cuelang.org/go/cue"
	"cuelang.org/go/cue/cuecontext"
	"cuelang.org/go/tools/flow"
)

func main() {
	ctx := cuecontext.New()
	f, err := os.ReadFile("x.cue")
	if err != nil {
		log.Fatal(err)
	}
	v := ctx.CompileBytes(f)
	if err := v.Err(); err != nil {
		log.Fatal(err)
	}

	controller := flow.New(&flow.Config{FindHiddenTasks: true}, v, ioTaskFunc)

	if err := controller.Run(context.Background()); err != nil {
		log.Fatal(err)
	}
}

func ioTaskFunc(v cue.Value) (flow.Runner, error) {
	inputPath := cue.ParsePath("$task")
	input := v.LookupPath(inputPath)

	if !input.Exists() {
		return nil, nil
	}

	return flow.RunnerFunc(func(t *flow.Task) error {
		return nil
	}), nil
}
-- x.cue --
package main

actions: {
	pre: #Run & {}

	issue_dep: [
		// here, the result is used not in other tasks.
		// it could be prepare data for other tasks.
		"\(pre.output)",
	]
}

#Run: {
	$task: "run"
	command?: [...string]
	output: string
}
```

### What did you expect to see?

task `actions.pre` execute success.

### What did you see instead?

got an error before task execute.
```shell
$ go run main.go
main.go:27: actions.issue_dep.0: invalid interpolation: non-concrete value string (type string) (and 1 more errors)
```

**Comment by rogpeppe** [comment]
I've bisected this regression to this CL: https://review.gerrithub.io/c/cue-lang/cue/+/1177330

**Comment by rogpeppe** [comment]
Firstly, thanks very much for the report and the reproducer. That's really useful!

Although this is technically a regression, ISTM that it may be because the above-mentioned fix actually exposes some genuine issues with your command flow. In particular, I'm not sure if it's allowed for non-task CUE code to depend of task-computed values, as you've got here. The [documentation says](https://pkg.go.dev/cuelang.org/go/tools/flow):

> Tasks may depend on other tasks. 

but it doesn't say that non-tasks may depend on tasks.

One workaround that you might be able to use here is to use a `let` expression rather than a field. For example, this adjustment to the code makes it work:

```
package main

actions: {
	pre: #Run & {}

	let issue_dep =  [
		// here, the result is used not in other tasks.
		// it could be prepare data for other tasks.
		"\(pre.output)",
	]
	post: #Run & {
		command: issue_dep
	}
}

#Run: {
	$task: "run"
	command?: [...string]
	output: string
}
```

If that doesn't work for you, perhaps you might be able to expand on what you're trying to do here to guide any potential fix?

**Comment by morlay** [comment]
@rogpeppe here is the real world usage.


https://github.com/octohelm/piper/blob/main/piper.cue#L33


We need fields to define task preset, which based on one or more seed tasks.


Example

https://github.com/octohelm/piper/blob/main/cuepkg/golang/project.cue

**Comment by rogpeppe** [comment]
Here is a test case that I think demonstrates the problem a little better:

```
exec go mod tidy
exec go run .
cmp stdout want-stdout
-- want-stdout --
final value: {
	param: "some output"
	t1: {
		$task:  true
		arg:    "some output"
		output: "some output"
	}
	t2: {
		arg: "some output"
	}
}
-- go.mod --
module test

go 1.22.1

require cuelang.org/go v0.8.0

-- main.go --
package main

import (
	"context"
	"fmt"
	"log"
	"os"

	"cuelang.org/go/cue"
	"cuelang.org/go/cue/cuecontext"
	"cuelang.org/go/tools/flow"
)

func main() {
	ctx := cuecontext.New()
	f, err := os.ReadFile("x.cue")
	if err != nil {
		log.Fatal(err)
	}
	v := ctx.CompileBytes(f)
	if err := v.Err(); err != nil {
		log.Fatal(err)
	}

	controller := flow.New(nil, v, func(v cue.Value) (flow.Runner, error) {
		if !v.LookupPath(cue.MakePath(cue.Str("$task"))).Exists() {
			return nil, nil
		}
		return flow.RunnerFunc(func(t *flow.Task) error {
			v := t.Value().FillPath(cue.MakePath(cue.Str("output")), "some output")
			t.Fill(v)
			return nil
		}), nil
	})

	if err := controller.Run(context.Background()); err != nil {
		log.Fatalf("controller run: %v", err)
	}
	final := controller.Value().LookupPath(cue.MakePath(cue.Str("b")))
	fmt.Printf("final value: %v\n", final)
}

-- x.cue --
#task: {
	$task: true
	arg?: string
	param?: string
	output?: string
}

a: #task
b: #tasks & {
	param: "\(a.output)"
}
// #tasks defines a collection of tasks, all of which share
// a common parameter that could be derived from the output
// of another task.
#tasks: {
	param: string
	t1: #task & {
		arg: param
	}
	t2: {
		arg: param
	}
}
```

This fails, but passes when the CUE version is set to `v0.7.1`.

**Comment by mvdan** [comment]
cc @nickfiggins who wrote the patch which introduced this regression - do you have any thoughts on Roger's example above, and what might be the best fix? For now we're thinking of doing a revert in v0.8.1 to un-break users like @morlay, keeping your added test case deactivated until we can figure out a better fix that keeps both test cases passing.

**Comment by nickfiggins** [comment]
It seems like one way to handle this is to only return certain errors (that are considered "permanent") instead of all errors from tasks. I put up a CL [here](https://review.gerrithub.io/c/cue-lang/cue/+/1185458) as a potential fix

**Comment by mvdan** [comment]
@rogpeppe I'm surprised that your revert did not fix and close this issue, as the original reproducer should work now. Are you leaving it open with the new title to track better errors? Can you please clarify?

**Comment by rogpeppe** [comment]
@mvdan I wasn't thinking clearly. Yes, I think this issue is now fixed - it's just the older issue (#1581) which should probably have been reopened.

**Comment by mvdan** [comment]
Thanks; restoring the old title for the sake of reference.

What older issue do you mean? #1571 is a discussion that's seemingly not related.

**Comment by rogpeppe** [comment]
> What older issue do you mean? #1571 is a discussion that's seemingly not related.

Oops, typo, I meant #1581. I've edited the comment now.
